dagster_and_r/R/iris.R

```diff
@@ -7,6 +8,36 @@ library(magrittr)
 reticulate::py_config()
 stopifnot(reticulate::py_module_available("dagster_pipes"))
 
+# Function to convert R types to Python types
+convert_r_to_python_types <- function(df) {
+    # Get R types
+    r_types <- sapply(df, class)
+    
+    # Define type mapping
+    type_mapping <- list(
+        "numeric" = "float",
+        "integer" = "int",
+        "character" = "str",
+        "factor" = "str",
+        "logical" = "bool",
+        "Date" = "datetime.date",
+        "POSIXct" = "datetime.datetime",
+        "POSIXlt" = "datetime.datetime"
+    )
+    
+    # Convert types
+    python_types <- sapply(r_types, function(x) {
+        if (x %in% names(type_mapping)) {
+            type_mapping[[x]]
+        } else {
+            "object"  # default type
+        }
+    })
+    
+    return(reticulate::r_to_py(as.list(python_types)))
+}
```

Comment:
It did the floats fine but defaults to "object" for the strings. I'll check into it some more

Reply:
Thanks for the info. I haven't had a chance to test your changes yet, but I'm hoping there's still a way that we can piggyback off reticulate's type conversion somehow.  

The [r_to_py.data.frame()](https://github.com/rstudio/reticulate/blob/61f0fa4873dc9d8ab6814ac474f332cdb6e8d489/R/conversion.R#L232-L260) function should be able to handle most R data types. Maybe we can pass `df` to this function, instead of using `sapply(df, class)`, and then extract the python types from the resulting data frame? Maybe reticulate is not recognizing that `pandas` is available, and `r_to_py_impl()` is called instead? 

I'll try taking a closer look in the next couple of days.

Reply:
Oh r_to_py.data.frame was the function I was missing after rifling through the reticulate docs. Yeah that might do it

Reply:
Leaning on `reticulate` more produces the below dictionary:

```R
reticulate::r_to_py(df)
{'Sepal.Length': dtype('float64'), 'Sepal.Width': dtype('float64'), 'Petal.Length': dtype('float64'), 'Petal.Width': dtype('float64'), 'Species': CategoricalDtype(categories=['setosa', 'versicolor', 'virginica'], ordered=False, categories_dtype=object)}
```

With a little more work, we can extract the name of each data type from its `dtype()` object:

```R
convert_r_to_python_types <- function(df) {
    df_pandas <- reticulate::r_to_py(df)
    dtypes_dict <- df_pandas$dtypes$to_dict()
    for (col_name in names(dtypes_dict)) {
        dtypes_dict[[col_name]] <- as.character(dtypes_dict[[col_name]]$name)
    }
    return(dtypes_dict)
}
```

This will materialize successfully, however, reticulate's `float64` is not recognized as a Python `float` by Dagster. Instead, it is interpreted as a `VARCHAR`:

![image](https://github.com/user-attachments/assets/2304c9d2-0628-4d27-b702-793ab8e02bcc)


I'm not sure if there's an implicit way to rely on reticulate here...

Reply:
I think relying on reticluate and then manually overriding a few might make the most sense (like float64 vs. float). 

I'll take the code here and add it to my PR with an override